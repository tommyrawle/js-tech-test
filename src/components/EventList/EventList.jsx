import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { formatDateOrTime } from '../../redux/utils';
import Market from '../Market/Market.jsx';
import * as actions from '../../redux/actions';
import { getDisplayableEvents } from '../../redux/selectors';
import { getAllLiveEvents, getMarketDetails } from '../../redux/getters';
import BGImage from '../../assets/bg.jpg';
import Spinner from '../Spinner/Spinner.jsx';
import {
  EventListTableContainer,
  EventListHeader,
  EventListTable,
  EventListTableRow,
  EventListTableCell,
  EventLink,
  EventImage,
  EventListInput
} from './EventList.styles.jsx';

export const EventList = ({ events, loading, setMarkets, setAllLiveEvents, setLoadingStatus, setError }) => {
  const [primaryMarkets, setPrimaryMarketsOption] = useState(false);

  const getEventListData = async primaryMarkets => {
    setLoadingStatus(true);
    const events = await getAllLiveEvents(primaryMarkets);
    setAllLiveEvents(events, primaryMarkets);
    const primaryMarketIds = events.reduce((arrayOfMarketIds, event) => {
      if (event.markets) {
        return [...arrayOfMarketIds, ...event.markets];
      } else return arrayOfMarketIds;
    }, []);
    if (primaryMarketIds.length) {
      Promise.all(primaryMarketIds.map(async market => await getMarketDetails(market)))
        .then(marketData => {
          setMarkets(marketData);
        })
        .catch(() => {
          let history = useHistory();
          setError({ code: 400, message: 'Could not process request' });
          history.push('/error');
        });
    }
  };

  useEffect(() => {
    getEventListData(primaryMarkets);
  }, [primaryMarkets]);

  return (
    <div>
      <EventImage src={BGImage} alt="live event" />
      {!loading ? (
        <EventListTableContainer>
          <EventListHeader>
            <h3>Live Events</h3>
            <div>
              <label>Show primary markets</label>
              <EventListInput
                type="checkbox"
                checked={primaryMarkets}
                onChange={e => setPrimaryMarketsOption(e.target.checked)}
              />
            </div>
          </EventListHeader>
          <EventListTable>
            <tbody>
              {events.map((event, i) => {
                return (
                  <EventListTableRow key={i}>
                    <EventListTableCell timeColumn>{formatDateOrTime(event.startTime, 'time')}</EventListTableCell>
                    <EventListTableCell eventCell>
                      <EventLink to={`/event/${event.eventId}`}>{event.name}</EventLink>
                    </EventListTableCell>
                    {event.markets && (
                      <EventListTableCell marketColumn>
                        <Market marketId={event.markets[0]} getOutcome />
                      </EventListTableCell>
                    )}
                  </EventListTableRow>
                );
              })}
            </tbody>
          </EventListTable>
        </EventListTableContainer>
      ) : (
        <Spinner />
      )}
    </div>
  );
};

EventList.propTypes = {
  events: PropTypes.arrayOf(PropTypes.object),
  loading: PropTypes.bool,
  setAllLiveEvents: PropTypes.func,
  setMarkets: PropTypes.func,
  setLoadingStatus: PropTypes.func,
  setError: PropTypes.func
};
const mapStateToProps = state => ({
  events: getDisplayableEvents(state),
  loading: state.loading
});
const mapDispatchToProps = dispatch => ({
  setAllLiveEvents: (arrayOfEvents, stillLoading) => dispatch(actions.setAllLiveEvents(arrayOfEvents, stillLoading)),
  setMarkets: arrayOfMarkets => dispatch(actions.setMarkets(arrayOfMarkets)),
  setLoadingStatus: bool => dispatch(actions.setLoadingStatus(bool)),
  setError: error => dispatch(actions.setError(error))
});

export default connect(mapStateToProps, mapDispatchToProps)(EventList);
