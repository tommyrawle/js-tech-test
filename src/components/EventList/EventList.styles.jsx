import styled from 'styled-components';

import { Link } from 'react-router-dom';

export const EventListInput = styled.input`
  margin-left: 5px;
`;
export const EventListTableContainer = styled.div`
  margin: 1rem auto;
  width: 70%;
  @media (max-width: 676px) {
    width: 100%;
  }
`;

export const EventListHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #002451;
  color: #ffffff;
  padding: 0 20px;
`;
export const EventListTable = styled.table`
  width: 100%;
`;

export const EventListTableRow = styled.tr`
  background-color: #f2f2f2;
  &:nth-child(even) {
    background-color: #ffffff;
    td {
      border-bottom: none;
    }
  }
`;

export const EventListTableCell = styled.td`
  padding: ${({ eventCell }) => (eventCell ? '10px 20px' : '')};
  border-bottom: 1px solid #ececec;
  ${({ timeColumn }) => timeColumn && 'width: 100px; text-align: center'};

  ${({ marketColumn }) => marketColumn && 'width: 40%; vertical-align: top'};
`;

export const EventLink = styled(Link)`
  text-decoration: none;
  color: inherit;
  &:hover {
    text-decoration: underline;
  }
`;

export const EventImage = styled.img`
  width: 100%;
  height: auto;
`;
