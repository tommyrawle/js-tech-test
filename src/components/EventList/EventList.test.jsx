import React from 'react';
import { shallow } from 'enzyme';
import { EventList } from './EventList.jsx';
import '@babel/polyfill';
describe('EventList component', () => {
  it('should render EventList component', () => {
    const mockProps = {
      events: [
        {
          eventId: 21249939,
          name: 'Shanghai Shenhua 0 v 0 Shandong Luneng Taishan',
          displayOrder: -1000,
          startTime: '2017-09-19T11:35:23.000Z'
        }
      ],
      loading: false,
      setMarkets: jest.fn(),
      setAllLiveEvents: jest.fn(),
      setLoadingStatus: jest.fn()
    };
    expect(shallow(<EventList {...mockProps} />)).toMatchSnapshot();
  });
});
