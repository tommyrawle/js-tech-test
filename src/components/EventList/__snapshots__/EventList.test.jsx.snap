// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventList component should render EventList component 1`] = `
<div>
  <styled.img
    alt="live event"
    src={null}
  />
  <styled.div>
    <styled.div>
      <h3>
        Live Events
      </h3>
      <div>
        <label>
          Show primary markets
        </label>
        <styled.input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </styled.div>
    <styled.table>
      <tbody>
        <styled.tr
          key="0"
        >
          <styled.td
            timeColumn={true}
          >
            12:35
          </styled.td>
          <styled.td
            eventCell={true}
          >
            <Styled(Link)
              to="/event/21249939"
            >
              Shanghai Shenhua 0 v 0 Shandong Luneng Taishan
            </Styled(Link)>
          </styled.td>
        </styled.tr>
      </tbody>
    </styled.table>
  </styled.div>
</div>
`;
